deps/NNNN-observability.md

```diff
@@ -0,0 +1,168 @@
+
+These requirements ensure that the observability framework in Dynamo is consistent, flexible, and easy to implement across all components, supporting the overall goal of enhanced fault tolerance and system reliability.
+
+
+# Proposal
+
+The proposal is to create a common library that allows each component/process to:
+
+* Expose metrics and/or health statistics on an HTTP endpoint.
+* Create an observability struct containing statistics (e.g., incr counter, gauge, and histogram).
+* Call a common API that mutates the observability struct.
+* Automatically export the observability struct to a Prometheus key-val format.
+
+Below is an ASCII diagram, showing 1) An observability struct containing pre-defined metric and tracing types, 2) common Observability APIs that take in the structs as parameters to mutate the struct (e.g., incr, gauge), and 3) A common HTTP library that also takes in the struct to expose an endpoint.
+
+```
++-------------------------------------+
+|          Observability              |
+|                 Struct              |
+|                                     |
+|   +-----------------------------+   |
+|   |       Metrics               |   |
+|   |   - counter                 |   |
+|   |   - gauge                   |   |
+|   |   - histogram               |   |
+|   +-----------------------------+   |
+|   |       Tracing               |   |
+|   |   - span                    |   |
+|   |   - context                 |   |
+|   +-----------------------------+   |
++-------------------------------------+
+       |                |                 |
+       |                |                 |
+       v                v                 v
++-------------+   +-------------+   +----------------------+
+|  Incr API   |   |  Gauge API  |   |  Common HTTP Library |
+|             |   |             |   |                      |
+|  - incr()   |   |  - set()    |   |  - expose endpoint() |
+|             |   |             |   |                      |
+|  +---------+|   |  +---------+|   |                      |
+|  | Call-   ||   |  | Call-   ||   |                      |
+|  | backs   ||   |  | backs   ||   |                      |
+|  +---------+|   |  +---------+|   |                      |
++-------------+   +-------------+   +----------------------+
+```
+In this diagram:
+* The `Observability Struct` encompasses both `Metrics` (like counter, gauge, histogram) and `Tracing` (like span, context).
+* The `Incr API` and `Gauge API` represent common Observability APIs that take the `Observability Struct` as parameters to modify it.
+* The `Common HTTP Library` also takes in the `Observability Struct` to expose an endpoint for monitoring or data collection.
+
+
+# Alternate Solutions
+
+## Alt 1: Use Particular Libraries
```

Comment:
I think we will use libraries - maybe this alternate should "use third party libraries directly"